File: src/whoami/collectors/grok.py
from __future__ import annotations

from datetime import date, datetime
from pathlib import Path

from whoami.collectors.jsonl_util import clip, iter_jsonl, on_day, parse_ts
from whoami.models import SessionEvent, TokenUsage
from whoami.paths import home


def collect(day: date) -> list[SessionEvent]:
    root = Path.home() / ".grok" / "sessions"
    env_home = home() / ".grok" / "sessions"
    roots = {root, env_home}
    sessions: list[SessionEvent] = []
    for base in roots:
        if not base.exists():
            continue
        for path in base.rglob("*.jsonl"):
            event = _parse_file(path, day)
            if event:
                sessions.append(event)
        for path in base.rglob("signals.json"):
            event = _parse_signals(path, day)
            if event:
                sessions.append(event)
    return _dedupe(sessions)


def _parse_file(path: Path, day: date) -> SessionEvent | None:
    tokens = TokenUsage()
    prompts: list[str] = []
    files: set[str] = set()
    tools: set[str] = set()
    times: list[datetime] = []
    model: str | None = None
    session_id = path.parent.name if path.name != path.parent.name else path.stem
    project = path.parent.parent.name

    relevant = False
    for obj in iter_jsonl(path):
        ts = parse_ts(obj.get("timestamp") or obj.get("createdAt") or obj.get("ts"))
        if ts and on_day(ts, day):
            relevant = True
            times.append(ts)

        model = obj.get("model") or model
        usage = obj.get("usage") or obj.get("token_usage") or obj.get("tokens")
        if isinstance(usage, dict):
            tokens = tokens.add(
                TokenUsage(
                    input_tokens=int(usage.get("input_tokens") or usage.get("input") or 0),
                    output_tokens=int(usage.get("output_tokens") or usage.get("output") or 0),
                    cache_read_tokens=int(usage.get("cache_read_tokens") or usage.get("cached") or 0),
                    cache_write_tokens=int(usage.get("cache_write_tokens") or 0),
                )
            )

        role = obj.get("role") or obj.get("type")
        text = obj.get("content") or obj.get("text") or obj.get("prompt")
        if role in {"user", "user_message", "prompt"} and text:
            rendered = text if isinstance(text, str) else str(text)
            if ts is None or on_day(ts, day):
                prompts.append(clip(rendered))

        tool = obj.get("tool") or obj.get("toolName") or obj.get("name")
        if isinstance(tool, str) and obj.get("type") in {"tool_call", "tool", "tool_use"}:
            tools.add(tool)

        for key in ("path", "file_path"):
            if isinstance(obj.get(key), str):
                files.add(obj[key])

    if not relevant and path.name not in {"chat_history.jsonl", "prompt_history.jsonl"}:
        return None
    if not relevant and not prompts:
        return None
    if not relevant:
        return None

    return SessionEvent(
        tool="grok-build",
        session_id=session_id,
        project=project,
        started_at=min(times) if times else None,
        ended_at=max(times) if times else None,
        model=model,
        tokens=tokens,
        user_prompts=prompts[:12],
        files_touched=sorted(files)[:30],
        tools_used=sorted(tools),
        title=prompts[0] if prompts else session_id,
    )


def _parse_signals(path: Path, day: date) -> SessionEvent | None:
    import json

    try:
        data = json.loads(path.read_text(encoding="utf-8"))
    except (OSError, json.JSONDecodeError):
        return None
    if not isinstance(data, dict):
        return None
    ts = parse_ts(data.get("updated_at") or data.get("timestamp"))
    if ts and not on_day(ts, day):
        return None
    usage = data.get("usage") if isinstance(data.get("usage"), dict) else {}
    tokens = TokenUsage(
        input_tokens=int(usage.get("input_tokens") or 0),
        output_tokens=int(usage.get("output_tokens") or 0),
    )
    if tokens.total == 0 and not data:
        return None
    return SessionEvent(
        tool="grok-build",
        session_id=path.parent.name,
        project=path.parent.parent.name,
        started_at=ts,
        ended_at=ts,
        tokens=tokens,
        title=str(data.get("title") or path.parent.name),
    )


def _dedupe(sessions: list[SessionEvent]) -> list[SessionEvent]:
    best: dict[str, SessionEvent] = {}
    for s in sessions:
        key = f"{s.tool}:{s.session_id}"
        prev = best.get(key)
        if not prev or s.tokens.total > prev.tokens.total:
            best[key] = s
    return list(best.values())

File: packages/core/src/collectors/grok.ts
import { existsSync, statSync } from "node:fs";
import { basename, dirname, join } from "node:path";
import { home } from "../paths";
import { rec, iterJsonl, walkFiles } from "../fsutil";
import { asText, clip, displayProject, intish, pickTitle } from "../text";
import { clusterTimestamps, onDay, parseTs, toIso } from "../time";
import { emptyTokens, type SessionEvent, type TokenUsage } from "../types";

function grokRoots(): string[] {
  return [join(home(), ".grok", "sessions")];
}

function isSessionDirName(name: string): boolean {
  return /^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$/i.test(name);
}

export async function collectGrok(day: string, timeZone: string): Promise<SessionEvent[]> {
  const sessions: SessionEvent[] = [];
  for (const root of grokRoots()) {
    if (!existsSync(root)) continue;
    const summaries = await walkFiles(root, (name) => name === "summary.json");
    const seen = new Set<string>();
    for (const summaryPath of summaries) {
      const dir = dirname(summaryPath);
      const sessionId = basename(dir);
      if (!isSessionDirName(sessionId) || seen.has(sessionId)) continue;
      seen.add(sessionId);
      const event = await parseSession(dir, sessionId, day, timeZone);
      if (event) sessions.push(event);
    }
    const updates = await walkFiles(root, (name) => name === "updates.jsonl");
    for (const path of updates) {
      const dir = dirname(path);
      const sessionId = basename(dir);
      if (!isSessionDirName(sessionId) || seen.has(sessionId)) continue;
      try {
        if (!onDay(statSync(path).mtime, day, timeZone)) continue;
      } catch {
        continue;
      }
      seen.add(sessionId);
      const event = await parseSession(dir, sessionId, day, timeZone);
      if (event) sessions.push(event);
    }
  }
  return sessions;
}

async function parseSession(
  dir: string,
  sessionId: string,
  day: string,
  timeZone: string,
): Promise<SessionEvent | null> {
  let title: string | null = null;
  let model: string | null = null;
  let cwd: string | null = null;
  const summaryPath = join(dir, "summary.json");
  if (existsSync(summaryPath)) {
    try {
      const data = rec(JSON.parse(await Bun.file(summaryPath).text()));
      if (data) {
        const info = rec(data.info);
        cwd = typeof info?.cwd === "string" ? info.cwd : null;
        if (typeof data.session_summary === "string" && data.session_summary.trim()) {
          title = clip(data.session_summary, 80);
        }
        if (typeof data.current_model_id === "string") model = data.current_model_id;
        const created = parseTs(data.created_at);
        const updated = parseTs(data.updated_at);
        const dayHit = onDay(created, day, timeZone) || onDay(updated, day, timeZone);
        if (!dayHit) return null;
      }
    } catch {
      /* fall through */
    }
  }

  const times: Date[] = [];
  const prompts: string[] = [];
  let maxTokens = 0;
  const updatesPath = join(dir, "updates.jsonl");
  if (existsSync(updatesPath)) {
    for await (const obj of iterJsonl(updatesPath)) {
      const ts = parseTs(obj.timestamp);
      const params = rec(obj.params);
      const update = rec(params?.update);
      const meta = rec(obj._meta) || rec(params?._meta) || rec(update?._meta);
      const tokenVal = intish(meta?.totalTokens ?? meta?.total_tokens);
      if (ts && onDay(ts, day, timeZone)) {
        times.push(ts);
        if (tokenVal > maxTokens) maxTokens = tokenVal;
        const kind = typeof update?.sessionUpdate === "string" ? update.sessionUpdate : "";
        if (kind === "user_message_chunk") {
          const text = asText(update?.content);
          if (text && !text.startsWith("<") && text.length < 400) prompts.push(clip(text));
        }
        const modelId = rec(update?._meta)?.modelId;
        if (typeof modelId === "string") model = modelId;
      }
    }
  }

  if (!times.length) return null;

  const clusters = clusterTimestamps(times);
  const hours = clusters.reduce((s, c) => s + c.hours, 0);
  const project = displayProject(cwd || basename(dirname(dir)));
  const tokens: TokenUsage = { ...emptyTokens(), input_tokens: maxTokens };
  return {
    tool: "grok-build",
    session_id: sessionId,
    project,
    started_at: toIso(clusters[0]?.start ?? null),
    ended_at: toIso(clusters.at(-1)?.end ?? null),
    model,
    tokens,
    user_prompts: prompts.slice(0, 12),
    files_touched: [],
    tools_used: [],
    title: title || pickTitle(prompts, sessionId),
    hours,
  };
}
